File: benchmarks/text/PathRepeatedSmallOps.c
#include <stdint.h>
#include <stddef.h>
#include <string.h>

struct path_facts {
    size_t length;
    size_t end;
    size_t segment_start;
    size_t extension_start;
    size_t directory_length;
    int has_extension;
};

static int is_separator(char value) {
    return value == '/';
}

static int join_path(char *destination, size_t capacity, const char *left, const char *right, size_t *length) {
    size_t left_length = strlen(left);
    size_t right_length = strlen(right);
    size_t right_start = 0;
    int insert_separator = 0;

    if (left_length > 0 && right_length > 0) {
        if (is_separator(left[left_length - 1])) {
            if (is_separator(right[0])) {
                right_start = 1;
            }
        } else if (!is_separator(right[0])) {
            insert_separator = 1;
        }
    }

    size_t right_copy_length = right_length - right_start;
    size_t total = left_length + right_copy_length + (insert_separator ? 1 : 0);
    if (total + 1 > capacity) {
        return 0;
    }

    size_t write = 0;
    memcpy(destination + write, left, left_length);
    write += left_length;
    if (insert_separator) {
        destination[write] = '/';
        write += 1;
    }

    memcpy(destination + write, right + right_start, right_copy_length);
    write += right_copy_length;
    destination[write] = '\0';
    *length = write;
    return 1;
}

static struct path_facts get_path_facts(const char *path) {
    size_t end = strlen(path);
    size_t length = end;
    while (end > 1 && is_separator(path[end - 1])) {
        end -= 1;
    }

    size_t separator = (size_t)-1;
    for (size_t index = end; index > 0; index -= 1) {
        if (is_separator(path[index - 1])) {
            separator = index - 1;
            break;
        }
    }

    size_t segment_start = separator == (size_t)-1 ? 0 : separator + 1;
    size_t directory_length = 0;
    if (separator == 0) {
        directory_length = 1;
    } else if (separator != (size_t)-1) {
        directory_length = separator;
    }

    size_t extension_start = end;
    int has_extension = 0;
    for (size_t index = end; index > segment_start + 1; index -= 1) {
        if (path[index - 1] == '.') {
            extension_start = index - 1;
            has_extension = 1;
            break;
        }
    }

    return (struct path_facts){
        .length = length,
        .end = end,
        .segment_start = segment_start,
        .extension_start = extension_start,
        .directory_length = directory_length,
        .has_extension = has_extension,
    };
}

static size_t extension_length(struct path_facts facts) {
    return facts.has_extension ? facts.end - facts.extension_start : 0;
}

static size_t base_name_length(struct path_facts facts) {
    if (facts.segment_start >= facts.end) {
        return 0;
    }

    return (facts.has_extension ? facts.extension_start : facts.end) - facts.segment_start;
}

int main(void) {
    char buffer[64] = {0};
    int64_t checksum = 0;

    for (int32_t i = 0; i < 5000; i += 1) {
        size_t length = 0;
        if (!join_path(buffer, sizeof(buffer), "alpha", "beta.txt", &length)) {
            return 1;
        }

        checksum += (int64_t)length;
        struct path_facts beta = get_path_facts("alpha/beta.txt");
        checksum += (int64_t)beta.length;
        checksum += (int64_t)extension_length(beta);
        checksum += (int64_t)base_name_length(beta);
        checksum += (int64_t)beta.directory_length;

        if (!join_path(buffer, sizeof(buffer), "alpha/beta", "gamma.txt", &length)) {
            return 2;
        }

        checksum += (int64_t)length;
        checksum += (int64_t)extension_length(get_path_facts("archive.tar.gz"));
    }

    return checksum == 320000 ? 0 : 3;
}

File: benchmarks/text/PathRepeatedSmallOps.rs
fn is_separator(value: u8) -> bool {
    value == b'/'
}

fn join_path(output: &mut String, left: &str, right: &str) -> bool {
    let left_bytes = left.as_bytes();
    let right_bytes = right.as_bytes();
    let mut right_start = 0_usize;
    let mut insert_separator = false;

    if !left_bytes.is_empty() && !right_bytes.is_empty() {
        if is_separator(left_bytes[left_bytes.len() - 1]) {
            if is_separator(right_bytes[0]) {
                right_start = 1;
            }
        } else if !is_separator(right_bytes[0]) {
            insert_separator = true;
        }
    }

    output.clear();
    if left.len() + right.len() + usize::from(insert_separator) > output.capacity() {
        return false;
    }

    output.push_str(left);
    if insert_separator {
        output.push('/');
    }
    output.push_str(&right[right_start..]);
    true
}

struct PathFacts<'a> {
    path: &'a str,
    end: usize,
    segment_start: usize,
    extension_start: usize,
    directory_length: usize,
    has_extension: bool,
}

impl<'a> PathFacts<'a> {
    fn new(path: &'a str) -> Self {
        let bytes = path.as_bytes();
        let mut end = bytes.len();
        while end > 1 && is_separator(bytes[end - 1]) {
            end -= 1;
        }

        let separator = (0..end).rev().find(|&index| is_separator(bytes[index]));
        let segment_start = separator.map_or(0, |index| index + 1);
        let directory_length = match separator {
            None => 0,
            Some(0) => 1,
            Some(index) => index,
        };

        let mut extension_start = end;
        let mut has_extension = false;
        for index in ((segment_start + 1)..end).rev() {
            if bytes[index] == b'.' {
                extension_start = index;
                has_extension = true;
                break;
            }
        }

        Self {
            path,
            end,
            segment_start,
            extension_start,
            directory_length,
            has_extension,
        }
    }

    fn path_len(&self) -> usize {
        self.path.len()
    }

    fn extension_len(&self) -> usize {
        if self.has_extension {
            self.end - self.extension_start
        } else {
            0
        }
    }

    fn base_name_len(&self) -> usize {
        if self.segment_start >= self.end {
            return 0;
        }

        let end = if self.has_extension {
            self.extension_start
        } else {
            self.end
        };
        end - self.segment_start
    }

    fn directory_name_len(&self) -> usize {
        self.directory_length
    }
}

fn main() {
    let mut output = String::with_capacity(64);
    let mut checksum = 0_i64;

    for _ in 0_i32..5000 {
        if !join_path(&mut output, "alpha", "beta.txt") {
            std::process::exit(1);
        }

        checksum += output.len() as i64;
        let beta = PathFacts::new("alpha/beta.txt");
        checksum += beta.path_len() as i64;
        checksum += beta.extension_len() as i64;
        checksum += beta.base_name_len() as i64;
        checksum += beta.directory_name_len() as i64;

        if !join_path(&mut output, "alpha/beta", "gamma.txt") {
            std::process::exit(2);
        }

        checksum += output.len() as i64;
        checksum += PathFacts::new("archive.tar.gz").extension_len() as i64;
    }

    if checksum != 320_000 {
        std::process::exit(3);
    }
}
